### PR TITLE
Feature/http

### PR DESCRIPTION
**Pull request by VitalySokolov**
Add work with real server.

### CODE REVIEW

File: src/app/shared/user.service.ts [hunk 1]
```diff
@@ -1,30 +1,40 @@
...
 import { AuthData } from './auth-data.model';
-import { Subject } from 'rxjs';
-import { Router } from '@angular/router';
+import { tap } from 'rxjs/operators';
 
 @Injectable({
   providedIn: 'root'
 })
 export class UserService {
+  readonly BASE_URL = 'http://localhost:3004';
```
Comment (vitaliy-bobrov) [inline src/app/shared/user.service.ts]: `private` as well

File: src/app/shared/user.service.ts [hunk 2]
```diff
@@ -1,30 +1,40 @@
...
-    this.user = {
-      name: user.name,
-      token: Math.round(Math.random() * 10000).toString()
-    };
-    this.authChange.next(true);
-    this.router.navigate(['/courses']);
+  public login(user: AuthData) {
+    return this.http.post<any>(`${this.BASE_URL}/auth/login`, {login: user.name, password: user.password})
+      .pipe(
+      tap((response) => {
```
Comment (vitaliy-bobrov) [inline src/app/shared/user.service.ts]: format

File: src/app/auth/login/login.component.ts [hunk 3]
```diff
@@ -9,23 +10,33 @@ import { UserService } from '@shared/user.service';
...
-  }
+    }
 
   onSubmit() {
     this.userService.login({
       name: this.loginForm.value.login,
       password: this.loginForm.value.password
-    });
+    })
+      .subscribe((response) => {
```
Comment (vitaliy-bobrov) [inline src/app/auth/login/login.component.ts]: need to be unsubscribed, use Subscription or takeUntil operator

Comment (vitaliy-bobrov) [reply src/app/auth/login/login.component.ts]: Unused param `response`

File: src/app/shared/user.service.ts [hunk 4]
```diff
@@ -1,30 +1,40 @@
 import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+import { HttpClient } from '@angular/common/http';
+
+import { pipe, Subject } from 'rxjs';
```
Comment (vitaliy-bobrov) [inline src/app/shared/user.service.ts]: where `pipe` was used?

File: src/app/shared/user.service.ts [hunk 5]
```diff
@@ -1,30 +1,40 @@
...
 
-  public login(user: AuthData): void {
-    this.user = {
-      name: user.name,
-      token: Math.round(Math.random() * 10000).toString()
-    };
-    this.authChange.next(true);
-    this.router.navigate(['/courses']);
+  public login(user: AuthData) {
+    return this.http.post<any>(`${this.BASE_URL}/auth/login`, {login: user.name, password: user.password})
```
Comment (vitaliy-bobrov) [inline src/app/shared/user.service.ts]: Add response type

File: src/app/video-courses/video-course.service.ts [hunk 6]
```diff
@@ -1,106 +1,101 @@
...
-  public addCourse(course: VideoCourseItem): VideoCourseItem {
-    const list = this.courseList;
-    course.id = list.length ? list[list.length - 1].id + 1 : 1;
-    this.courseList = [...list, course];
+  public getCourseListChange(): Subject<VideoCourseItem[]> {
+    return this.courseListChange;
+  }
+
+  public getVideoCourses(): void {
+    if (this.courseList.length > 0) {
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/video-course.service.ts]: `if (this.courseList.length)`

File: src/app/video-courses/video-course.service.ts [hunk 7]
```diff
@@ -1,106 +1,101 @@
...
+    } else {
+      this.fetchVideoCourses();
+    }
+  }
 
-    return course;
+  private fetchVideoCourses() {
+    this.httpClient.get<CourseItem[]>(`${this.BASE_URL}`)
+      .subscribe(response => {
+        this.courseList = response.map(course => {
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/video-course.service.ts]: this.courseList = response.map(this.convertToVideoCourseItem);

File: src/app/video-courses/video-course.service.ts [hunk 8]
```diff
@@ -1,106 +1,101 @@
...
   }
 
   public getCourse(id: number): VideoCourseItem | undefined {
     return this.courseList.find((course) => course.id === id);
   }
 
   public removeCourse(id: number): void {
+    this.httpClient.delete(`${this.BASE_URL}/${id}`)
+      .subscribe();
     this.courseList = this.courseList.filter((course) => course.id !== id);
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/video-course.service.ts]: Should be done only on success

File: src/app/video-courses/video-course.service.ts [hunk 9]
```diff
@@ -1,106 +1,101 @@
...
+      .subscribe((updatedCourseItem) => {
+          const updatedCourse = this.convertToVideoCourseItem(updatedCourseItem);
+
+          this.courseList = this.courseList.map((course) => {
+            return course.id === updatedCourse.id ? updatedCourse : course;
+          });
+          this.courseListChange.next(this.courseList);
+        },
+        (error) => {
+          this.courseList = this.courseList.map((course) => {
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/video-course.service.ts]: No changes needed on an error, just show something to the user.

File: src/app/video-courses/video-course.service.ts [hunk 10]
```diff
@@ -1,106 +1,101 @@
...
+          this.courseListChange.next(this.courseList);
+        });
+  }
+
+  private convertToVideoCourseItem(course: CourseItem): VideoCourseItem {
+    return {
+      id: course.id,
+      title: course.name,
+      description: course.description,
+      authors: course.authors.slice(),
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/video-course.service.ts]: Do we need to make a copy?

File: src/app/video-courses/video-course.service.ts [hunk 11]
```diff
@@ -1,106 +1,101 @@
...
+      });
   }
 
   public getCourse(id: number): VideoCourseItem | undefined {
     return this.courseList.find((course) => course.id === id);
   }
 
   public removeCourse(id: number): void {
+    this.httpClient.delete(`${this.BASE_URL}/${id}`)
+      .subscribe();
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/video-course.service.ts]: error handling

File: src/app/auth/login/login.component.ts (hunk 3 above)
Comment (VitalySokolov) [reply src/app/auth/login/login.component.ts]: But it is an http request. Should I unsubscribe from it anyway?

Unused param removed.

File: src/app/shared/user.service.ts (hunk 4 above)
Comment (VitalySokolov) [reply src/app/shared/user.service.ts]: It is a leftover. Removed. Thanks.

File: src/app/shared/user.service.ts (hunk 5 above)
Comment (VitalySokolov) [reply src/app/shared/user.service.ts]: Done.

File: src/app/video-courses/video-course.service.ts (hunk 6 above)
Comment (VitalySokolov) [reply src/app/video-courses/video-course.service.ts]: Done.

File: src/app/video-courses/video-course.service.ts (hunk 7 above)
Comment (VitalySokolov) [reply src/app/video-courses/video-course.service.ts]: Done.

File: src/app/video-courses/video-course.service.ts (hunk 10 above)
Comment (VitalySokolov) [reply src/app/video-courses/video-course.service.ts]: You are right. Thanks.

**Pull request merged**